libfreefare/mifare_desfire.c

```diff
@@ -141,49 +145,89 @@ static ssize_t	 read_data(FreefareTag tag, uint8_t command, uint8_t file_no, off
+
+	if (!msg) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	memset (msg_buf, 0, sizeof(msg_buf));
+	msg_buf[0] = 0x90; // CLA
+	msg_buf[1] = msg[0]; // INS
+	// P1 & P2 is zero by default
+
+	if (msg_len > 1) {
+	    len += msg_len;
+	    msg_buf[4] = msg_len - 1; // Lc
+	    memcpy (msg_buf + 5, msg + 1, msg_len - 1);
+	}
+
+	MIFARE_DESFIRE (tag)->last_picc_error = OPERATION_OK;
+	MIFARE_DESFIRE (tag)->last_pcd_error = OPERATION_OK;
+
+	DEBUG_XFER (msg_buf, len, "===> ");
+
+	if ((rc = nfc_initiator_transceive_bytes (tag->device, msg_buf, len, res_buf, sizeof(res_buf), 500)) < 2) {
+		errno = (errno == ETIMEDOUT) ? errno : EIO;
+	    return -1;
+	}
+
+	DEBUG_XFER (res_buf, rc, "<=== ");
+
+	rc--;
+	if (rc > res_size) {
+		errno = ENOBUFS;
+		return -1;
+	}
+
+    int lpe = res_buf[rc];
+
+	res[rc - 1] = lpe;
+
+	if ((1 == rc) && (ADDITIONAL_FRAME != res_buf[rc]) && (OPERATION_OK != res_buf[rc])) {
+
+	    if (AUTHENTICATION_ERROR == lpe)
+			errno = EACCES;
+
+		MIFARE_DESFIRE (tag)->last_picc_error = lpe;
+		return -1;
+	}
+
+	if (res_len)
+		*res_len = rc;
+	memcpy (res, res_buf, rc - 1);
+
+	errno = 0;
+	return 0;
+}
+
+
+/*
+ * Convenience macros.
  */
-#define DESFIRE_TRANSCEIVE2(tag, msg, msg_len, res) \
-    do { \
-	DEBUG_FUNCTION(); \
-	static uint8_t __msg[MAX_CAPDU_SIZE + 5] = { 0x90, 0x00, 0x00, 0x00, 0x00, /* ..., */ 0x00 }; \
-	/*                                       CLA   INS   P1    P2    Lc    PAYLOAD    LE*/ \
-	static uint8_t __res[MAX_RAPDU_SIZE + 1]; \
-	size_t __len = 5; \
-	errno = 0; \
-	if (!msg) return errno = EINVAL, -1; \
-	__msg[1] = msg[0]; \
-	if (msg_len > 1) { \
-	    __len += msg_len; \
-	    __msg[4] = msg_len - 1; \
-	    memcpy (__msg + 5, msg + 1, msg_len - 1); \
-	} \
-	/* reply length */ \
-	__msg[__len-1] = 0x00; \
-	MIFARE_DESFIRE (tag)->last_picc_error = OPERATION_OK; \
-	MIFARE_DESFIRE (tag)->last_pcd_error = OPERATION_OK; \
-	DEBUG_XFER (__msg, __len, "===> "); \
-	int _res; \
-	if ((_res = nfc_initiator_transceive_bytes (tag->device, __msg, __len, __res, __##res##_size + 1, 500)) < 0) { \
-	    return errno = (errno == ETIMEDOUT) ? errno : EIO, -1; \
-	} \
-	__##res##_n = _res; \
-	DEBUG_XFER (__res, __##res##_n, "<=== "); \
-	res[__##res##_n-2] = __res[__##res##_n-1]; \
-	__##res##_n--; \
-	if ((1 == __##res##_n) && (ADDITIONAL_FRAME != res[__##res##_n-1]) && (OPERATION_OK != res[__##res##_n-1])) { \
-	    if (res[0] == AUTHENTICATION_ERROR) \
-		errno = EACCES; \
-	    return MIFARE_DESFIRE (tag)->last_picc_error = res[0], -1; \
-	} \
-	memcpy (res, __res, __##res##_n - 1); \
-    } while (0)
 
-
+#ifdef WITH_DEBUG
+	// define a small but cute wrapper to allow debug output
+	#define mifare_desfire_transceive(tag, msg, msg_len, res, res_size, res_len) ({ \
+		int __err = 0; \
+	    do { \
+		DEBUG_FUNCTION(); \
+		__err = desfire_transceive(tag, msg, msg_len, res, res_size, res_len); \
+	    } while (0); \
+	    __err; \
+	  })
+#else
+	#define mifare_desfire_transceive desfire_transceive
+#endif
```

Comment:
For the sake of simplicity, add the debug code directly to the `desfire_transceive()` function:

```c
static int
desfire_transceive(FreefareTag tag, const char *msg, size_t msg_len, char *res, size_t res_size, size_t *res_len)
{
#ifdef WITH_DEBUG
    DEBUG_FUNCTION();
#endif
    // more code
}
```

Reply:
Sorry, won't work, been there ;)
Doing so will show always "desfire_transceive" in the debug output.
The compiler does not support a macro showing the calling function.
IMHO the script is mandarory. And it works

Reply:
Ah!  Didn't checked what the macro does :unamused: Yeah, tracing things in C is a PITA.